### PR TITLE
fix(common): returnable argument to `.then()`

### LINKED ISSUE #13258: `future<T>::then()` cannot return the provided future
This code does not work:

```cc
#include <google/cloud/future.h>

future<int> F(future<int> f) {
  return f.then([](auto g) { return g; });
}
```

Though there is no reason why it shouldn't. It is easy to workaround the problem:

```cc
#include <google/cloud/future.h>

future<int> F(future<int> f) {
  return f.then([](auto g) { return g.get(); });
}
```

### LINKED ISSUE #13314: Propagate abandoned future errors via `future<T>::then()`
Consider this code:

```cc
std::future<int> F() { return std::promise<int>().get_future(); }
google::cloud::future<int> G() { return google::cloud::promise<int>().get_future(); }
```

The expression `F().get()` will throw a `std::future_error(std::future_errc::broken_promise)`[^1]:

https://en.cppreference.com/w/cpp/thread/promise/%7Epromise

The expression `G().get()` behaves the same way, that all sounds good.  Now consider this function:

```cc
google::cloud::future<int> H() {
  return google::cloud::promise<int>().get_future().then([](auto g) { return 2 * g.get(); });
}
```

Currently `H().get()` deadlocks.  That is not a good thing.  I can fix it, the fix is trivial, but unfortunately we have unit tests that depend on the existing behavior.

We should fix the tests, unfortunately it is not that easy.  These tests (and the code they are testing) needs to work with exceptions disabled (i.e., the `noex` builds aka `-fno-exceptions`).  With exceptions disabled `.get()` terminates the program if the promise was abandoned (or the `future<T>` contains any exception).

So we need to find what `promise` is being destroyed and abandoning the future, and most likely, set the value to some kind of `google::cloud::Status` before destroying the promise.

[^1]: this is the only sensible choice.  Once cannot leave `.get()` blocked forever, and there is no valid `int` value to return.

### PR DESCRIPTION
**Pull request by coryan**
The main motivation for these changes is to support:

```cc
future<T> F(future<T> f) {
  return f.then([](auto g) { /* stuff */; return g; });
}
```

This is common enough, and was broken for fairly obscure reasons.  I completely changed the implementation of `.then()`. It was too hacky to do something else, and I think the new implementation is simpler too.

Fixes #13314  and fixes #13258 

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13316)

### CODE REVIEW

**Comment by coryan** [comment]
The `check-api` results are false positives.

1. Evidently we no longer create an instantiation of `future<future<future<void>>>::get()`, we do not need to, but if anybody creates an instance it will work.
```
Step #3: Removed Symbols  1 
Step #3: 
Step #3: ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Step #3: future_generic.h
Step #3: namespace google::cloud::v2_20
Step #3: future<future<future<void> > >::get ( )
Step #3: _ZN6google5cloud5v2_206futureINS2_INS2_IvEEEEE3getEv
```

2. This is an `absl::typedef` that went away:
```
Step #3: Problems with Data Types, Low Severity  1 
Step #3: 
Step #3: ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Step #3: type_traits.h
Step #3: namespace absl
Step #3: [+] typedef remove_reference_t  1 
```

**Comment by coryan** [comment]
The warnings on `Debian:bookworm` (GCC 12 FWIW) look like false positives:

```
/usr/include/c++/12/bits/exception_ptr.h:117:33: error: '*(std::__exception_ptr::exception_ptr*)((char*)&<unnamed> + offsetof(absl::debian3::variant<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>,absl::debian3::variant<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantCopyAssignBaseNontrivial<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantMoveAssignBaseNontrivial<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantCopyBaseNontrivial<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantMoveBaseNontrivial<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantStateBaseDestructorNontrivial<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::<unnamed>.absl::debian3::variant_internal::VariantStateBase<absl::debian3::monostate, std::__exception_ptr::exception_ptr, google::cloud::v2_20::internal::FutureVoid, google::cloud::v2_20::internal::FutureValueRetrieved>::state_)).std::__exception_ptr::exception_ptr::_M_exception_object' may be used uninitialized [-Werror=maybe-uninitialized]
  117 |       : _M_exception_object(__o._M_exception_object)
      |                             ~~~~^~~~~~~~~~~~~~~~~~~
/workspace/google/cloud/internal/future_impl.h: In member function 'void google::cloud::v2_20::internal::AsyncConnectionReadyFuture::RunIteration()':
/workspace/google/cloud/internal/future_impl.h:176:65: note: '<anonymous>' declared here
  176 |   future_shared_state() : future_shared_state([] {}, ValueType{}) {}
      |                                                                 ^
cc1plus: all warnings being treated as errors

```

First, it is a `maybe-unitialized`. Second, it is in `absl::variant<>` (or `std::exception_ptr`) we have no control over either and are supposed to initialize properly.  Third, we use the version of Abseil included in this distro, which is fairly old.

I am going to suppress the warning using `#pragma`.

File: google/cloud/internal/future_impl.h [hunk 1]
```diff
@@ -53,7 +107,7 @@ struct IsReferenceWrapper<std::reference_wrapper<U>> : public std::true_type {};
  * Continuations (the parameters to a `.then()` call) can be of arbitrary
  * types: any callable (lambdas, function pointers, `std::function<>`) should be
  * accepted. We want to hold these continuations as type-erased objects, so we
- * can call them without having to know their types.
+ * can call them without having to their types.
```
Comment (devbww) [inline google/cloud/internal/future_impl.h:110]: Accidental edit?

File: google/cloud/internal/future_impl.h [hunk 2]
```diff
@@ -67,42 +121,105 @@ class Continuation {  // NOLINT(readability-identifier-naming)
...
  *
  * The shared state contains either (1) the value stored by the promise, or (2)
- * the exception stored by the promise, or (3) a continuation to be called when
- * the shared state is satisfied.
+ * the exception stored by the promise, (3) nothing, if the future has not been
+ * satisfied, (4) a different kind of nothing, if the future was satisfied and
+ * the value or exception was extracted.  It may also contain a continuation to
+ * be called when the shared state is satisfied.
  *
  * The shared state value can be retrieved only once, this is enforced by
```
Comment (devbww) [inline google/cloud/internal/future_impl.h:143]: s/once, this/once. This/

File: google/cloud/internal/future_impl.h [hunk 3]
```diff
@@ -67,42 +121,105 @@ class Continuation {  // NOLINT(readability-identifier-naming)
...
+        ThrowDelegate(
+            std::move(ex),
+            "future<T>::get() had an exception but exceptions are disabled");
+      }
+      T operator()(FutureValueRetrieved) {
+        ThrowFutureError(std::future_errc::no_state, "future<T>::get()");
+      }
+      T operator()(absl::monostate) {
+        ThrowFutureError(std::future_errc::no_state, "future<T>::get()");
+      }
```
Comment (devbww) [inline google/cloud/internal/future_impl.h:215]: Should we distinguish the two messages?

File: google/cloud/internal/future_impl.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/future_impl.h]: Fixed

File: google/cloud/internal/future_impl.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/future_impl.h]: Fixed

File: google/cloud/internal/future_impl.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/future_impl.h]: Done.

File: google/cloud/internal/future_impl.h [hunk 4]
```diff
@@ -53,7 +107,7 @@ struct IsReferenceWrapper<std::reference_wrapper<U>> : public std::true_type {};
  * Continuations (the parameters to a `.then()` call) can be of arbitrary
  * types: any callable (lambdas, function pointers, `std::function<>`) should be
  * accepted. We want to hold these continuations as type-erased objects, so we
- * can call them without having to know their types.
+ * can call them without having to known their types.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/future_impl.h]: D'oh, fixed.

**Pull request merged**